wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/InternalEnvironmentFactory.java

```diff
@@ -0,0 +1,129 @@
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.workspace.server.spi;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+import org.eclipse.che.api.core.ValidationException;
+import org.eclipse.che.api.core.model.workspace.Warning;
+import org.eclipse.che.api.core.model.workspace.config.Environment;
+import org.eclipse.che.api.core.model.workspace.config.MachineConfig;
+import org.eclipse.che.api.core.model.workspace.config.ServerConfig;
+import org.eclipse.che.api.installer.server.InstallerRegistry;
+import org.eclipse.che.api.installer.server.exception.InstallerException;
+import org.eclipse.che.api.installer.shared.model.Installer;
+import org.eclipse.che.api.workspace.server.model.impl.EnvironmentImpl;
+import org.eclipse.che.api.workspace.server.model.impl.ServerConfigImpl;
+
+/**
+ * Factory for Environment specific internal representation Related but not really bound to some
+ * specific Infrastructure to let Infrastructure apply multiple different implementations, some of
+ * which can be considered as a "native format", while others as rather "supported, adopted formats"
+ *
+ * <p>Expected to be bound as a MapBinder with unique String as a key, like: MapBinder<String,
+ * InternalEnvironmentFactory> environmentFactories = MapBinder.newMapBinder(binder(), String.class,
+ * InternalEnvironmentFactory.class);
+ * environmentFactories.addBinding("uniq_name").to(SubclassOfInternalEnvironmentFactory.class);
+ *
+ * @author gazarenkov
+ */
+public abstract class InternalEnvironmentFactory {
+
+  protected final InstallerRegistry installerRegistry;
+  protected final RecipeRetriever recipeRetriever;
+
+  public InternalEnvironmentFactory(
+      InstallerRegistry installerRegistry, RecipeRetriever recipeRetriever) {
+    this.installerRegistry = installerRegistry;
+    this.recipeRetriever = recipeRetriever;
+  }
+
+  /**
+   * validates internals of Environment and creates instance of InternalEnvironment
+   *
+   * @param environment the environment
+   * @return InternalEnvironment
+   * @throws InfrastructureException if infrastructure specific error occures
+   * @throws ValidationException if validation fails
+   */
+  public final InternalEnvironment create(final Environment environment)
+      throws InfrastructureException, ValidationException {
+
+    Map<String, InternalMachineConfig> machines = new HashMap<>();
+    List<Warning> warnings = new ArrayList<>();
+
+    // Workaround related to written docker image id into location instead of content
+    EnvironmentImpl envCopy = new EnvironmentImpl(environment);
+    if ("dockerimage".equals(environment.getRecipe().getType())
```

Comment:
I think this logic should be moved to `DockerimageInternalEnvironmentFactory`

Reply:
wanted to have create(final Environment environment) final but OK, will change, not a big deal to me